app/uk/gov/hmrc/economiccrimelevyreturns/services/EclLiabilityService.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.economiccrimelevyreturns.services
+
+import play.api.mvc.RequestHeader
+import uk.gov.hmrc.economiccrimelevyreturns.connectors.EclReturnsConnector
+import uk.gov.hmrc.economiccrimelevyreturns.models.EclReturn
+import uk.gov.hmrc.economiccrimelevyreturns.utils.EclTaxYear
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendHeaderCarrierProvider
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class EclLiabilityService @Inject() (eclReturnsConnector: EclReturnsConnector)(implicit
+  ec: ExecutionContext
+) extends FrontendHeaderCarrierProvider {
+
+  def calculateLiability(eclReturn: EclReturn)(implicit request: RequestHeader): Future[EclReturn] =
+    (
+      eclReturn.relevantAp12Months,
+      eclReturn.relevantApLength,
+      eclReturn.relevantApRevenue,
+      eclReturn.carriedOutAmlRegulatedActivityForFullFy,
+      eclReturn.amlRegulatedActivityLength
+    ) match {
+      case (Some(true), _, Some(relevantApRevenue), Some(true), _)                                 =>
+        calculateLiabilityImpl(EclTaxYear.YearInDays, relevantApRevenue, EclTaxYear.YearInDays, eclReturn)
+      case (Some(false), Some(relevantApLength), Some(relevantApRevenue), Some(true), _)           =>
+        calculateLiabilityImpl(relevantApLength, relevantApRevenue, EclTaxYear.YearInDays, eclReturn)
+      case (Some(true), _, Some(relevantApRevenue), Some(false), Some(amlRegulatedActivityLength)) =>
+        calculateLiabilityImpl(EclTaxYear.YearInDays, relevantApRevenue, amlRegulatedActivityLength, eclReturn)
+      case (
+            Some(false),
+            Some(relevantApLength),
+            Some(relevantApRevenue),
+            Some(false),
+            Some(amlRegulatedActivityLength)
+          ) =>
+        calculateLiabilityImpl(relevantApLength, relevantApRevenue, amlRegulatedActivityLength, eclReturn)
+      case _                                                                                       =>
+        throw new IllegalStateException("AP and/or AML data not found in ECL return")
+    }
```

Comment:
You could change this to return `Option[Future[EclReturn]]`

and then implement it like:

```
for {
      _                 <- eclReturn.relevantAp12Months
      relevantApRevenue <- eclReturn.relevantApRevenue
      _                 <- eclReturn.carriedOutAmlRegulatedActivityForFullFy
    } yield calculateLiabilityImpl(
      relevantApLength = eclReturn.relevantApLength.getOrElse(EclTaxYear.YearInDays),
      relevantApRevenue = relevantApRevenue,
      amlRegulatedActivityLength = eclReturn.amlRegulatedActivityLength.getOrElse(EclTaxYear.YearInDays),
      eclReturn = eclReturn
    )
```

and then handle the `Option` in the navigators instead of throwing the exception

Reply:
I do actually mean `Option[Future[EclReturn]]` in this case

We only call the API if the necessary data items exist, if they don't we don't call the API and return none, therefore in this case the Future is 'optional'